numpy/core/overrides.py

```diff
@@ -66,29 +66,36 @@
     """)
 
 
-ArgSpec = collections.namedtuple('ArgSpec', 'args varargs keywords defaults')
+# Avoid extra getattr calls in verify_matching_signatures
+_Parameter_empty = inspect.Parameter.empty
 
 
 def verify_matching_signatures(implementation, dispatcher):
     """Verify that a dispatcher function has the right signature."""
-    implementation_spec = ArgSpec(*getargspec(implementation))
-    dispatcher_spec = ArgSpec(*getargspec(dispatcher))
-
-    if (implementation_spec.args != dispatcher_spec.args or
-            implementation_spec.varargs != dispatcher_spec.varargs or
-            implementation_spec.keywords != dispatcher_spec.keywords or
-            (bool(implementation_spec.defaults) !=
-             bool(dispatcher_spec.defaults)) or
-            (implementation_spec.defaults is not None and
-             len(implementation_spec.defaults) !=
-             len(dispatcher_spec.defaults))):
-        raise RuntimeError('implementation and dispatcher for %s have '
-                           'different function signatures' % implementation)
-
-    if implementation_spec.defaults is not None:
-        if dispatcher_spec.defaults != (None,) * len(dispatcher_spec.defaults):
-            raise RuntimeError('dispatcher functions can only use None for '
-                               'default argument values')
+    implementation_sig = inspect.signature(implementation)
```

Comment:
I can only assume the timings are bad because `inspect.signature` is slower than `getargspec`

Reply:
Yes, some (very rough) profiling indicates that this is the case. Using the following setup:

```python
from numpy.core.overrides import verify_matching_signatures                     
                                                                                
def foo(                                                                        
    a, b, m, n, k, copy=False, color="red", normed=False, weights=None,         
    bins=10, density=None, size=1024, shape=(3, 4)                              
):                                                                              
    pass                                                                        
                                                                                
def _foo_dispatcher(                                                            
    a, b, m, n, k, copy=None, color=None, normed=None, weights=None,            
    bins=None, density=None, size=None, shape=None                              
):                                                                              
    pass
```

`timeit` gives:

On 6e35c2a5a6 (this PR):
```
In [2]: %timeit verify_matching_signatures(foo, _foo_dispatcher)                             
73.1 µs ± 879 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

On 2283e2659 (root of branch):
```
In [2]: %timeit verify_matching_signatures(foo, _foo_dispatcher)                             
4.85 µs ± 11.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

So the implementation based on `inspect.signature` is about 15x slower than the original based on `_inspect.getargspec`.

One idea I had was to try to re-implement this with `inspect.getfullargspec`, which is the more full-featured extension of the original (now deprecated) `inspect.getargspec`. I will update with `-X importtime` & profiling info from the attempt.

Reply:
> One idea I had was to try to re-implement this with inspect.getfullargspec, which is the more full-featured extension of the original (now deprecated) inspect.getargspec.

I'm pretty sure this is implemented in terms of `inspect.signature` - at any rate, it obtains an `inspect.Signature` object then  does some extra processing.